google/cloud/spanner_dbapi/checksum.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2020 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+"""API to calculate checksums of SQL statements results."""
+
+import hashlib
+import pickle
+
+
+class ResultsChecksum:
+    """Cumulative checksum.
+
+    Used to calculate a total checksum of all the results
+    returned by operations executed within transaction.
+    Includes methods for checksums comparison.
+    These checksums are used while retrying an aborted
+    transaction to check if the results of a retried transaction
+    are equal to the results of the original transaction.
+    """
+
+    def __init__(self):
+        self.checksum = hashlib.sha256()
+        self.count = 0  # counter of consumed results
+
+    def __len__(self):
+        """Return the number of consumed results.
+
+        :rtype: :class:`int`
+        :returns: The number of results.
+        """
+        return self.count
+
+    def __eq__(self, other):
+        """Check if checksums are equal.
+
+        :type other: :class:`google.cloud.spanner_dbapi.checksum.ResultsChecksum`
+        :param other: Another checksum to compare with this one.
+        """
+        return self.checksum.digest() == other.checksum.digest()
+
+    def consume_result(self, result):
+        """Add the given result into the checksum.
+
+        :type result: Union[int, list]
+        :param result: Streamed row or row count from an UPDATE operation.
+        """
+        self.checksum.update(pickle.dumps(result))
+        self.count += 1
+
+
+def _compare_checksums(original, retried):
+    """Compare the given checksums.
+
+    Raise an error if the given checksums have consumed
+    the same number of results, but are not equal.
+
+    :type original: :class:`~google.cloud.spanner_v1.transaction.ResultsChecksum`
+    :param original: results checksum of the original transaction.
+
+    :type retried: :class:`~google.cloud.spanner_v1.transaction.ResultsChecksum`
+    :param retried: results checksum of the retried transaction.
+
+    :raises: :exc:`RuntimeError` in case if checksums are not equal.
+    """
+    if original is not None:
+        if len(retried) == len(original) and retried != original:
```

Comment:
I think there's a misunderstanding regarding why we must keep track of `count` here. During a transaction retry we need to execute all statements again and check the results of the retried statements with the original statements. For queries there are two possibilities during a retry:
1. The results of a query have already been fully consumed by the client application (i.e. all rows have been fetched). In that case, during a retry, we must re-execute the statement and fetch all rows. Once we have done that, we can compare the checksums with each other. The `count` should also be equal for both result sets, but we do not necessarily need to check that, as if they were to be different, the checksums will also be different.
1. The results of a query have NOT yet been fully consumed by the client application (i.e. an `Aborted` error occurred while the client application was iterating over the result set). If that happens, we need to re-execute the statement and then advance the underlying iterator of the result set `count` steps and compare the checksums with each other. If the checksums are equal, the results that the client application has seen is equal to the results that we got during the retry, and the transaction may proceed. We also need to replace the underlying iterator of the result set that has been returned to the client application with the iterator that we got during the retry, as that iterator is now associated with the current transaction, while the old iterator is no longer valid. In the Java implementation, replacing the underlying iterator (`delegateResultSet`) [happens here](https://github.com/googleapis/java-spanner/blob/eddd6ad4e5093ee21290b85f15fa432d071bae59/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ChecksumResultSet.java#L177).

In the Java implementation the step that is described

Reply:
We're keeping counts inside checksum objects. When checksums are compared, they are first compared by length. If lengths are different, nothing will happen, as we consider this like: "the retried transaction is not yet came to the same point" or "the retried transaction have more results - it passed the fail point of the original transaction". But if the checksums have the same length and different checksums, it'll fail.